linera-views/src/map_view.rs

```diff
@@ -258,40 +310,53 @@ where
     ///   let mut map = ByteMapView::load(context).await.unwrap();
     ///   map.insert(vec![0,1], String::from("Hello"));
     ///   map.insert(vec![1,2], String::from("Bonjour"));
-    ///   map.insert(vec![2,2], String::from("Bonjour"));
+    ///   map.insert(vec![1,3], String::from("Bonjour"));
+    ///   let prefix = vec![1];
     ///   let mut count = 0;
     ///   map.for_each_key_while(|_key| {
     ///     count += 1;
-    ///     Ok(count < 1)
-    ///   }).await.unwrap();
-    ///   assert_eq!(count, 1);
+    ///     Ok(count < 3)
+    ///   }, prefix).await.unwrap();
+    ///   assert_eq!(count, 2);
     /// # })
     /// ```
-    pub async fn for_each_key_while<F>(&self, mut f: F) -> Result<(), ViewError>
+    pub async fn for_each_key_while<F>(&self, mut f: F, prefix: Vec<u8>) -> Result<(), ViewError>
     where
         F: FnMut(&[u8]) -> Result<bool, ViewError> + Send,
     {
-        let mut updates = self.updates.iter();
+        let prefix_len = prefix.len();
+        let mut updates = self.updates.range(get_interval(prefix.clone()));
         let mut update = updates.next();
+        let mut deleted_prefixes = BTreeSet::new();
+        let mut lower_bound = if prefix_len == 0 {
+            NextLowerKeyIterator::new(&self.deleted_prefixes)
+        } else {
+            for deleted_prefix in &self.deleted_prefixes {
+                if deleted_prefix.len() >= prefix_len && deleted_prefix[0..prefix_len] == prefix {
+                    deleted_prefixes.insert(deleted_prefix[prefix_len..].to_vec());
+                }
+            }
+            NextLowerKeyIterator::new(&deleted_prefixes)
```

Comment:
If `NextLowerKeyIterator` took an `Iterator` instead of a `BTreeSet`, we could probably just do something like:

```suggestion
            NextLowerKeyIterator::new(self.deleted_prefixes.range(get_interval(prefix.clone())))
```

and avoid iterating over the whole map, and comparing each entry.

Reply:
It is not as simple since the prefixes need to be shortened. But I will try.

Reply:
Yes, but at least in my limited understanding of rust, I could not see how to get away with creating a new `BTreeSet`.

So, instead I went into some kind of goose chase of restructuring the code. Maybe it was overkill, but at least it was successful.